src/index.ts

```diff
@@ -66,24 +69,47 @@ export function PluginPure(options: PureAnnotationsOptions): Plugin {
               && node.callee.type === 'Identifier'
               && functionSet.has(node.callee.name)
             ) {
-              for (const _comment of ast.comments || /* v8-ignore */ []) {
-                const comment = withLocations(_comment)
-                if (comment.end <= node.start && comment.value.includes('__PURE__')) {
-                  return
-                }
-              }
-              s.prependRight(node.start, '/*@__PURE__*/ ')
+              const annotation = '/*@__PURE__*/ '
+              s.prependRight(node.start, annotation)
+              annotations.push({ start: node.start, annotation })
             }
           },
         })
 
-        if (s.hasChanged()) {
+        if (annotations.length) {
+          fixAst(ast, annotations)
+
           return {
             code: s.toString(),
+            ast: withLocations(ast),
             map: options.sourcemap ? /* v8-ignore */ s.generateMap({ hires: true }) : undefined,
           }
         }
       },
     },
   }
 }
+
+function fixAst(ast: Program, annotations: Annotation[]) {
```

Comment:
is there any reason we can't augment the ast in one pass?

Reply:
~~No, we can't.
For example, the Program node has { start: 0, end: 100, ...} we have to update the end according to how many annotations we added in total. But before we completely walked all nodes once, we didn't know the count. And this happens to all other nodes, and both start and end.
That's why we have to do the calculation in the second walk.~~

Yes, we can 🚀